projects/charger/src/charger_controller.c

```diff
@@ -0,0 +1,127 @@
+#include "charger_controller.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "can_interval.h"
+#include "can_transmit.h"
+#include "charger_events.h"
+#include "event_queue.h"
+#include "generic_can.h"
+#include "generic_can_msg.h"
+#include "status.h"
+
+#define CHARGER_PERIOD_US 1000000  // 1 Second as defined in datasheet.
+
+#define CHARGER_EXPECTED_RX_DLC 5
+#define CHARGER_EXPECTED_RX_ID 0x18FF50E5
+
+#define CHARGER_EXPECTED_TX_DLC 5
+#define CHARGER_EXPECTED_TX_ID 0x1806E5F4
+
+static CanInterval *s_interval;
+static ChargerStatus *s_charger_status;
+static GenericCanMsg s_tx_msg;
+
+typedef union J1939CanId {
+  uint32_t raw_id;
+  struct {
+    uint32_t source_address : 8;  // Source
+    uint32_t pdu_specifics : 8;   // Destination
+    uint32_t pdu_format : 8;      // Packet Format
+    uint32_t dp : 1;              // Always 0
+    uint32_t r : 1;               // Always 0
+    uint32_t priority : 3;        // Anything
+  };
+} J1939CanId;
```

Comment:
Technically doesn't meet our style guide? I'd suggest making a header for all the unions and structs that are needed to convert between their CAN protocol and ours. (ex. `ChargerControllerTxData`)

Reply:
Done.